vllm/worker/multi_step_model_runner.py

```diff
@@ -614,34 +614,66 @@ def _pythonize_sampler_output(
 
     frozen_model_input = model_input.frozen_model_input
     assert frozen_model_input.sampling_metadata is not None
+    sampling_metadata = frozen_model_input.sampling_metadata
     # samples generation should have been skipped
     assert not output.outputs
 
     pinned_buffer = pinned_sampled_token_buffer[:model_input.num_queries]
 
-    # CPU GPU sync
-    pinned_buffer = pinned_buffer.copy_(sampled_token_ids, non_blocking=False)
+    # We are guaranteed output tensors are ready, so it is safe to
+    # pythonize the sampler output & obtain CPU-side logprobs.
+    #
+    # However we should check whether logprobs pythonization may
+    # be skipped entirely, i.e. because no logprobs were requested
+    # or pythonization was not deferred. To that end,
+    #
+    # * `prompt_logprobs_are_requested_for_prefill` signals that
+    #   there are *any* prefill-phase requests which specify that
+    #   prompt logprobs should be returned.
+    #
+    # * `any_logprobs_are_requested` signals that there are any
+    #   requests which (1) specify that sample logprobs should be
+    #   returned, or (2) are in the prefill phase AND specify that
+    #   prompt logprobs should be returned.
+    #
+    # Later on, these flags cause adjustments to the pythonization
+    # process to accommodate logprobs.
+
+    seq_groups = sampling_metadata.seq_groups
+    prompt_logprobs_are_requested_for_prefill = any([
+        ((sg.sampling_params.prompt_logprobs is not None) and sg.is_prompt)
+        for sg in seq_groups
+    ])
```

Comment:
We will only have one sequence in prefill group.

```suggestion
    prompt_logprobs_are_requested_for_prefill = seq_groups[0].is_prompt and seq_groups[0].sampling_params.prompt_logprobs is not None
```

Reply:
there cannot be multi SequenceGroup's that are in the prefill stage?

Reply:
Ah sorry it can be. I messed up with the parentheses and thought it was sequences inside the sequence groups. Please ignore this then. Meanwhile we could make it a bit more clean:

```suggestion
    prompt_logprobs_are_requested_for_prefill = any([
        sg.sampling_params.prompt_logprobs is not None and sg.is_prompt
        for sg in seq_groups
    ])
```